### PR TITLE
chatgpt

### PR DESCRIPTION
**Pull request by adrelanos**
Changes for the dev branch as suggested by chatgpt AI.

Could you please check if any or even all of these changes are sensible?

Untested.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -82,16 +82,18 @@ long current_time_ms(void) {
 }
 
 long random_between(long lower, long upper) {
-        // default to max if the interval is not valids
+        // default to max if the interval is not valid
         if (lower >= upper)
                 return upper;
 
-        return lower + randombytes_uniform(upper+1);
+        return lower + randombytes_uniform(upper - lower + 1);
```
Comment (vmonaco) [inline src/main.c:89]: This looks like the only change to program logic, which seems like a bug fix. Good catch

File: lib/systemd/system/kloak.service [hunk 2]
```diff
@@ -55,7 +55,7 @@ NoNewPrivileges=true
 RestrictRealtime=true
 RestrictNamespaces=true
 SystemCallArchitectures=native
-SystemCallFilter=ioctl nanosleep select write read openat close brk fstat lseek mmap mprotect munmap rt_sigaction rt_sigprocmask access execve getuid arch_prctl set_tid_address set_robust_list prlimit64 pread64 getrandom
+SystemCallFilter=ioctl nanosleep select write read openat close brk fstat lseek mmap mprotect munmap rt_sigaction rt_sigprocmask access execve getuid arch_prctl set_tid_address set_robust_list prlimit64 pread64 getrandom newfstatat clock_nanosleep pselect6
```
Comment (vmonaco) [inline lib/systemd/system/kloak.service:58]: I wonder if newfstatat is actually needed. stat() is actually called anywhere in main.c, right?

**Comment by vmonaco** [comment]
I think all the changes make sense, maybe just double check that the additional SystemCallFilter args are needed. These might have gotten picked up from eventcap.c which isn't really needed to run the service.

**Review by vmonaco — APPROVED** [review]

**Comment by adrelanos** [comment]
> maybe just double check that the additional SystemCallFilter args are needed

:heavy_check_mark: Done.

These are needed. I reproduced needing them independently. So first @hardcore-sushi figured out that these are needed and sent https://github.com/vmonaco/kloak/pull/40. It was my mistake to not check this upstream repository first so I re-invented these fixes downstream.

In summary, @hardcore-sushi and me came up with the same needed SystemCallFilter seccomp. (Just in a different order but that doesn't matter at all.)

File: src/main.c (hunk 1 above)
Comment (adrelanos) [reply src/main.c:89]: Nice!

File: lib/systemd/system/kloak.service (hunk 2 above)
Comment (adrelanos) [reply lib/systemd/system/kloak.service:58]: I have no idea why it is needed. On Debian bullseye it wasn't needed. But on Debian bookworm it's needed. I can only speculate that one of the underlying libraries changed and now requires that syscall.

**Comment by adrelanos** [comment]
Added more commits.

* fixed Apparmor
* disabled seccomp due to https://github.com/vmonaco/kloak/pull/38#issuecomment-1627688486

**Comment by adrelanos** [comment]
* This is a merge request for the `dev` branch.
* kloak is functional in so far that:
  * The kloak systemd unit is running.
  * No apparmor denied messages.
  * No seccomp issues.
* What I didn't do: Test if kloak dev branch actually functional. I suggest that I create 2 issues:
  * merge dev branch into master branch when ready
  * dev branch functionality testing
* I think this is good enough for a `dev` branch.
* Good to process/merge/close all the pull requests.

Therefore, if it is a reasonable request, I kindly request to to merge this pull request. This is useful to have a `dev` branch which is easily re-viewable by others.

For any remaining TODO, I suggest to keep using github issues.

**Comment by adrelanos** [comment]
* Also tested changing the systemd unit from `kloak` to `kloak -v` (verbose mode). It showed that all keyboard press events and all mouse movements are recognized by kloak.

**Comment by adrelanos** [comment]
I guess https://github.com/vmonaco/kloak/pull/50 is a more significant contribution. In case of any merge conflicts, the chatgpt improvements could be easily re-done.

**Comment by skyzzuu** [comment]
@adrelanos  Looking at the changes, I think they should merge pretty well since I don't think any common lines in main.c were changed. I think the only conflict that would come up would be the SystemCallFilter line in the systemd unit file, but I can just update mine to be the union of the system calls we both added to the filter line so all of them are in there.

File: src/main.c [hunk 3]
```diff
@@ -82,16 +82,18 @@ long current_time_ms(void) {
...
 }
 
-void set_rescue_keys(char* rescue_keys_str) {
-        char *_rescue_keys_str = malloc(strlen(rescue_keys_str) + 1);
+void set_rescue_keys(const char* rescue_keys_str) {
+        char* _rescue_keys_str = malloc(strlen(rescue_keys_str) + 1);
         strncpy(_rescue_keys_str, rescue_keys_str, strlen(rescue_keys_str));
+        _rescue_keys_str[strlen(rescue_keys_str)] = '\0';
```
Comment (adrelanos) [inline src/main.c:95]: Please check if this line makes sense.

Comment (vmonaco) [reply src/main.c:95]: yes null char should be copied over

**Review by vmonaco — APPROVED** [review]
tested and everything works as expected

**Pull request merged**